### PR TITLE
add unit tests and get_model

### PR DESCRIPTION
**Pull request by liopeer**
## What has changed and why?
- add get_model to FeatureExtractor interface
- implement get_model method for all models

## How has it been tested?
- added unit tests for all models and also added some other missing unit tests for torchvision

## Did you update [CHANGELOG.md](../CHANGELOG.md)?

- [ ] Yes
- [x] Not needed (internal change)

## Did you update the documentation?

- [ ] Yes
- [x] Not needed (internal change without effects for user)

### CODE REVIEW

File: src/lightly_train/_callbacks/export.py [hunk 1]
```diff
@@ -39,7 +39,7 @@ def on_train_epoch_end(self, trainer: Trainer, pl_module: LightningModule) -> No
             export_path = self._out_dir / "exported_last.pt"
             if export_path.exists():
                 export_path.unlink()
-
```
Comment (liopeer) [inline src/lightly_train/_callbacks/export.py]: formatting

File: src/lightly_train/_methods/dinov2/dinov2_transform.py [hunk 2]
```diff
@@ -33,6 +33,7 @@ class DINOv2LocalViewTransformArgs(DINOLocalViewTransformArgs):
         default_factory=DINOv2LocalViewRandomResizeArgs
     )
 
+
```
Comment (liopeer) [inline src/lightly_train/_methods/dinov2/dinov2_transform.py:36]: formatting

File: tests/_models/super_gradients/test_segmentation_module.py [hunk 3]
```diff
@@ -36,3 +36,15 @@ def test_forward_features(self) -> None:
         x = torch.rand(1, 3, 224, 224)
         out = fe.forward_features(x)["features"]
         assert out.shape == (1, 1024, 7, 7)
+
+    def test_forward_pool(self) -> None:
```
Comment (liopeer) [inline tests/_models/super_gradients/test_segmentation_module.py:40]: added to make test suite complete

File: src/lightly_train/_models/rfdetr/rfdetr.py [hunk 4]
```diff
@@ -37,7 +37,7 @@ def __init__(self, model: Module) -> None:
         feature_dim = encoder._out_feature_channels[-1]
         assert isinstance(feature_dim, int)
 
-        self._model = model
+        self._model: Module = model
```
Comment (stegmuel) [inline src/lightly_train/_models/rfdetr/rfdetr.py]: If only `model.backbone.encoder` is trained why not use `[model]`here?

File: src/lightly_train/_models/rfdetr/rfdetr.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ def __init__(self, model: Module) -> None:
...
         # Set model to training mode. This is necessary for RFDETR pretrained
         # models as the DINOv2 backbone is in eval mode by default.
         self._model.train()
```
Comment (stegmuel) [inline src/lightly_train/_models/rfdetr/rfdetr.py]: Here, it is probably not needed to set the whole `_model` in train mode.

File: src/lightly_train/_models/rfdetr/rfdetr.py [hunk 6]
```diff
@@ -55,3 +55,6 @@ def forward_features(self, x: Tensor) -> ForwardFeaturesOutput:
 
     def forward_pool(self, x: ForwardFeaturesOutput) -> ForwardPoolOutput:
         return {"pooled_features": self._pool(x["features"])}
+
+    def get_model(self) -> Module:
+        return self._model
```
Comment (stegmuel) [inline src/lightly_train/_models/rfdetr/rfdetr.py]: If you change in the `__init__` you'll also need to change here.

File: src/lightly_train/_models/super_gradients/customizable_detector.py [hunk 7]
```diff
@@ -59,3 +59,6 @@ def forward_features(self, x: Tensor) -> ForwardFeaturesOutput:
 
     def forward_pool(self, x: ForwardFeaturesOutput) -> ForwardPoolOutput:
         return {"pooled_features": self._pool(x["features"])}
+
+    def get_model(self) -> Module:
+        return self._model
```
Comment (stegmuel) [inline src/lightly_train/_models/super_gradients/customizable_detector.py]: Shouldn't `[model]` also be used in this class? Only `model.backbone` seems to be trained.

File: src/lightly_train/_models/super_gradients/segmentation_module.py [hunk 8]
```diff
@@ -61,3 +61,6 @@ def forward_features(self, x: Tensor) -> ForwardFeaturesOutput:
 
     def forward_pool(self, x: ForwardFeaturesOutput) -> ForwardPoolOutput:
         return {"pooled_features": self._pool(x["features"])}
+
+    def get_model(self) -> Module:
+        return self._model
```
Comment (stegmuel) [inline src/lightly_train/_models/super_gradients/segmentation_module.py]: Shouldn't [model] also be used in this class? Only model.backbone seems to be trained.

File: src/lightly_train/_models/ultralytics/ultralytics.py [hunk 9]
```diff
@@ -69,6 +69,9 @@ def forward_features(self, x: Tensor) -> ForwardFeaturesOutput:
     def forward_pool(self, x: ForwardFeaturesOutput) -> ForwardPoolOutput:
         return {"pooled_features": self._pool(x["features"])}
 
+    def get_model(self) -> YOLO:
+        return self._model
```
Comment (stegmuel) [inline src/lightly_train/_models/ultralytics/ultralytics.py]: Shouldn't [model] also be used in this class? Only model.backbone seems to be trained.

**Review by stegmuel — COMMENTED** [review]
Left some comments, which all boil down to the same concern/question.

File: src/lightly_train/_callbacks/export.py (hunk 1 above)
Comment (guarin) [reply src/lightly_train/_callbacks/export.py]: Looks like ruff changed behavior, are you on the latest ruff version?

Comment (liopeer) [reply src/lightly_train/_callbacks/export.py]: Yes. I am on 11.8 and that's how it chose to format this file.

File: src/lightly_train/_models/rfdetr/rfdetr.py (hunk 4 above)
Comment (liopeer) [reply src/lightly_train/_models/rfdetr/rfdetr.py:40]: Did not want to change the implementation. However it might indeed make sense to change it. Will do it.

Comment (liopeer) [reply src/lightly_train/_models/rfdetr/rfdetr.py]: Did now. I initially just wanted to change as little as possible to the original implementation.

File: src/lightly_train/_models/rfdetr/rfdetr.py (hunk 5 above)
Comment (liopeer) [reply src/lightly_train/_models/rfdetr/rfdetr.py]: Also changed.

File: src/lightly_train/_models/rfdetr/rfdetr.py (hunk 6 above)
Comment (liopeer) [reply src/lightly_train/_models/rfdetr/rfdetr.py]: Also changed.

File: src/lightly_train/_models/super_gradients/customizable_detector.py (hunk 7 above)
Comment (liopeer) [reply src/lightly_train/_models/super_gradients/customizable_detector.py]: Changed.

File: src/lightly_train/_models/super_gradients/segmentation_module.py (hunk 8 above)
Comment (liopeer) [reply src/lightly_train/_models/super_gradients/segmentation_module.py]: Changed.

File: src/lightly_train/_models/ultralytics/ultralytics.py (hunk 9 above)
Comment (liopeer) [reply src/lightly_train/_models/ultralytics/ultralytics.py]: Changed, but had to delete an existing unit test (since the DFL layer is now not listed anymore in `named_parameters()`

File: src/lightly_train/_models/super_gradients/customizable_detector.py [hunk 10]
```diff
@@ -46,7 +46,7 @@ def supported_model_classes(cls) -> tuple[type[Module], ...]:
         return cls._SUPPORTED_MODEL_CLASSES
 
     def feature_dim(self) -> int:
-        out_channels: list[int] | int = self._model.backbone.out_channels
+        out_channels: list[int] | int = self._model[0].backbone.out_channels
```
Comment (stegmuel) [inline src/lightly_train/_models/super_gradients/customizable_detector.py]: Why not:
`out_channels: list[int] | int = self._.backbone.out_channels`

File: src/lightly_train/_models/super_gradients/segmentation_module.py [hunk 11]
```diff
@@ -47,7 +47,7 @@ def supported_model_classes(cls) -> tuple[type[Module], ...]:
 
     def feature_dim(self) -> int:
         out_channels: list[int] | int = (
-            self._model.backbone.get_backbone_output_number_of_channels()
+            self._model[0].backbone.get_backbone_output_number_of_channels()
```
Comment (stegmuel) [inline src/lightly_train/_models/super_gradients/segmentation_module.py]: Why not:
out_channels: list[int] | int = self._.backbone.out_channels

**Review by stegmuel — APPROVED** [review]
Approved with two minor comments.
